### PR TITLE
Doc/xhr hydration examples

### PR DESCRIPTION
**Pull request by hchevalier**
https://github.com/shakacode/react_on_rails/issues/921#issuecomment-392229113

---
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1095)

### CODE REVIEW

File: spec/dummy/app/views/pages/xhr_refresh.html.erb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    <%= submit_tag 'Refresh', id: 'refresh', name: 'refresh' %>
+  <% end %>
+</div>
+<hr/>
+
+<h1>React Rails Client Rehydration</h1>
+<p>
+  This example demonstrates client side manual rehydration after a component replacement through XHR.<br/><br/>
+
+  The "Refresh" button on this page will trigger an asynchrounous refresh of component-container content.<br/>
```
Comment (hchevalier) [inline spec/dummy/app/views/pages/xhr_refresh.html.erb]: Typo on "asynchronous"

File: spec/dummy/app/views/pages/xhr_refresh.html.erb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  Components will be prerendered by the server and inserted in the DOM (spec/dummy/app/views/pages/xhr_refresh.js.erb)<br/>
+  No client rehydration will occur, preventing any event handler to be correctly attached<br/><br/>
+
+  Thus, the onChange handler of the HelloWorld component won't trigger whereas the one from HellowWorldRehydratable will, thanks to the "hydrate" javscript event dispacthed from xhr_refresh.js.erb<br />
```
Comment (hchevalier) [inline spec/dummy/app/views/pages/xhr_refresh.html.erb]: Typo on "javascript"

**Comment by hchevalier** [comment]
Review status: 9 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[README.md, line 546 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LD__JcdD4k8Hr-ROk_-:-LDbOSlC4C3_QAy7qR7P:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/788bc8a97ab9203507e4c1b93356907749da7de5/README.md#L546)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

red dots mean extra spaces
</blockquote></details>

Done.

---

*[spec/dummy/app/views/pages/xhr_refresh.html.erb, line 2 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LD__es4CpeRH-cYWrcJ:-LDbORNsAVLeBRMrSxHr:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/788bc8a97ab9203507e4c1b93356907749da7de5/spec/dummy/app/views/pages/xhr_refresh.html.erb#L2)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

This should not be needed if we're using the rails/webpacker way.

The jquery bits should be put into the webpacker tag.

This is pre-webpacker: https://github.com/shakacode/react_on_rails/blob/master/spec/dummy_no_webpacker/app/views/layouts/application.html.erb#L19

This is post-webpacker: 

https://github.com/shakacode/react_on_rails/blob/master/spec/dummy/app/views/layouts/application.html.erb#L16

this is the post webpacker jquery: https://github.com/shakacode/react_on_rails/blob/master/spec/dummy/client/package.json#L27,

https://github.com/shakacode/react_on_rails/blob/master/spec/dummy/client/webpack.client.rails.build.config.js#L106,

https://github.com/shakacode/react_on_rails/blob/master/spec/dummy/client/webpack.client.base.config.js#L65
</blockquote></details>

Done.

---

*[spec/dummy/app/assets/javascripts/jquery_rails_manifest.js, line 2 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LD__LtB5zK5xi98Nzv8:-LDbT7rf51OmGXLt4ztj:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/788bc8a97ab9203507e4c1b93356907749da7de5/spec/dummy/app/assets/javascripts/jquery_rails_manifest.js#L2)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

do we really need to use sprockets for jquery?
</blockquote></details>

Done.

---

*[spec/dummy/client/app/components/HelloWorldRehydratable.js, line 48 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LD_cRYR3IDkucYdsg-A:-LDbOLKk5VdCl8mGEN82:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/788bc8a97ab9203507e4c1b93356907749da7de5/spec/dummy/client/app/components/HelloWorldRehydratable.js#L48)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

get the script tag from just querying for the script tag with the where the `<script data-dom-id>`  matches the component's div id.

definitely make sure all this JS passes the linter.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1095)*

**Comment by justin808** [comment]
Some suggestions. Looking good.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[spec/dummy/client/app/startup/clientRegistration.jsx, line 4 at r3](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDiY8cI-zbnypmzzQGg:-LDiY8cI-zbnypmzzQGh:b-7dy69) ([raw file](https://github.com/shakacode/react_on_rails/blob/a23d3142b0b232dfc794e4c01c70ef1f182e22da/spec/dummy/client/app/startup/clientRegistration.jsx#L4)):*
> ```JSX
> import 'es5-shim';
> import 'jquery'
> import {} from 'jquery-ujs'
> ```

Any explanation for the addition of these?

I think we should use something like this for all of these:

* https://github.com/shakacode/react-webpack-rails-tutorial/search?q=es5-shim&unscoped_q=es5-shim
* https://github.com/shakacode/react-webpack-rails-tutorial/search?q=jquery-ujs&unscoped_q=jquery-ujs

---

*[spec/dummy/client/app/components/HelloWorldRehydratable.jsx, line 47 at r3](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDiXyHK4KI4pshlMdJB:-LDiXyHLDnKDIwQVQgPC:b-98330f) ([raw file](https://github.com/shakacode/react_on_rails/blob/a23d3142b0b232dfc794e4c01c70ef1f182e22da/spec/dummy/client/app/components/HelloWorldRehydratable.jsx#L47)):*
> ```JSX
>       // Get component specification <script> tag
>       const componentSpecificationTag = document.querySelector(`script[data-dom-id=${component.id}]`);
>       if (componentSpecificationTag) {
> ```

How can this be falsey for a React on Rails component?

---


*Comments from [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-:-LDiYp9q3ee9awEJDSgR:bq0g4ka)*

**Comment by hchevalier** [comment]
Review status: 10 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[spec/dummy/client/app/startup/clientRegistration.jsx, line 4 at r3](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDiY8cI-zbnypmzzQGg:-LDmoxbu6Nah65cW5NaM:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/a23d3142b0b232dfc794e4c01c70ef1f182e22da/spec/dummy/client/app/startup/clientRegistration.jsx#L4)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Any explanation for the addition of these?

I think we should use something like this for all of these:

* https://github.com/shakacode/react-webpack-rails-tutorial/search?q=es5-shim&unscoped_q=es5-shim
* https://github.com/shakacode/react-webpack-rails-tutorial/search?q=jquery-ujs&unscoped_q=jquery-ujs
</blockquote></details>

Done.

---

*[spec/dummy/client/app/components/HelloWorldRehydratable.jsx, line 47 at r3](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDiXyHK4KI4pshlMdJB:-LDmoyqYDm299iBqUgn0:b-896fix) ([raw file](https://github.com/shakacode/react_on_rails/blob/a23d3142b0b232dfc794e4c01c70ef1f182e22da/spec/dummy/client/app/components/HelloWorldRehydratable.jsx#L47)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

How can this be falsey for a React on Rails component?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1095)*

**Comment by justin808** [comment]
Just one thing I see...

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[spec/dummy/client/webpack.client.base.config.js, line 73 at r4](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDnTG2i-stRJO2PBhqy:-LDnTG2i-stRJO2PBhqz:b-ksfltr) ([raw file](https://github.com/shakacode/react_on_rails/blob/ffdb42ea0e6e3485755d17fe3cf4c0c1f494b2d5/spec/dummy/client/webpack.client.base.config.js#L73)):*
> ```JavaScript
>           {
>             loader: 'expose-loader',
>             options: '$'
> ```

Why do we need to use $ for jQuery in example? I don't like it, as it the `$` is used for ES6 string interpolation.

If you just want to let others know about this option, you can add a comment to the line above for jQuery to create a similar entry to expose `$` if your rails JS code needs $ exposed globally.

---


*Comments from [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-:-LDnTpTs1SCPOgR_EFgQ:b-plytdf)*

**Comment by justin808** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[README.md, line 545 at r5](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDsxGAE8_L7XCiyDWNa:-LDsxGAE8_L7XCiyDWNb:b-1fqnkg) ([raw file](https://github.com/shakacode/react_on_rails/blob/f33fda0b1139448079da1c37d6ed9f6cdd433cf4/README.md#L545)):*
> ```Markdown
> 
>   Note: client hydration will not trigger for components rendered through XHR. You will have to handle it with javascript.
>   For an example, see [spec/dummy/app/views/pages/xhr_refresh.rb](https://github.com/shakacode/react_on_rails/tree/master/spec/dummy/app/views/pages/xhr_refresh.rb). 
> ```

Are you sure this link works? This will be done from the README.md so it might not.

---

*[spec/dummy/client/app/components/HelloWorldRehydratable.jsx, line 43 at r5](https://reviewable.io/reviews/shakacode/react_on_rails/1095#-LDsxlwZ1FzKzC429xVp:-LDsxlwZ1FzKzC429xVq:bb74pdg) ([raw file](https://github.com/shakacode/react_on_rails/blob/f33fda0b1139448079da1c37d6ed9f6cdd433cf4/spec/dummy/client/app/components/HelloWorldRehydratable.jsx#L43)):*
> ```JSX
>     const match = document.querySelectorAll(`[id^=${registeredComponentName}-react-component-]`);
>     // Not all browsers support forEach on NodeList so we go with a classic for-loop
>     for (let i = 0; i < match.length; i += 1) {
> ```

The 

---


*Comments from [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1095)*

**Comment by justin808** [comment]
Thanks @hchevalier!

**Pull request merged**